src/sources/apache_metrics/parser.rs

```diff
@@ -0,0 +1,919 @@
+                Ok(value) => Some(Ok(StatusFieldStatistic::ConnsTotal(value))),
+                Err(err) => Some(Err(ParseError {
+                    key: key.to_string(),
+                    err: err.into(),
+                })),
+            },
+            "ConnsAsyncWriting" => match value.parse::<f64>() {
+                Ok(value) => Some(Ok(StatusFieldStatistic::ConnsAsyncWriting(value))),
+                Err(err) => Some(Err(ParseError {
+                    key: key.to_string(),
+                    err: err.into(),
+                })),
+            },
+            "ConnsAsyncClosing" => match value.parse::<f64>() {
+                Ok(value) => Some(Ok(StatusFieldStatistic::ConnsAsyncClosing(value))),
+                Err(err) => Some(Err(ParseError {
+                    key: key.to_string(),
+                    err: err.into(),
+                })),
+            },
+            "ConnsAsyncKeepAlive" => match value.parse::<f64>() {
+                Ok(value) => Some(Ok(StatusFieldStatistic::ConnsAsyncKeepAlive(value))),
+                Err(err) => Some(Err(ParseError {
+                    key: key.to_string(),
+                    err: err.into(),
+                })),
+            },
+            "Scoreboard" => Some(Ok(StatusFieldStatistic::Scoreboard(value))),
+
+            _ => None,
+        }
+    }
+}
+
+/// Parses the text output from Apache's mod_status and returns:
+///
+/// - A list of metrics generated from the output
+/// - A list of parse errors that were encountered
+///
+/// # Arguments
+///
+/// - `payload` - the mod_status output
+/// - `namespace` - the namespace to put the generated metrics in
+/// - `now` - the time the payload was fetched
+/// - `tags` - any base tags to apply to the metrics
+pub fn parse(
+    payload: &str,
+    namespace: &str,
+    now: DateTime<Utc>,
+    tags: Option<&BTreeMap<String, String>>,
+) -> impl Iterator<Item = Result<Metric, ParseError>> {
+    // We use a HashMap rather than a Vector as mod_status has
+    // BusyWorkers/IdleWorkers repeated
+    // https://bz.apache.org/bugzilla/show_bug.cgi?id=63300
+    let parsed = payload
+        .lines()
+        .into_iter()
+        .filter_map(|l| {
+            let mut parts = l.splitn(2, ":");
+            let key = parts.next();
+            let value = parts.next().map(|s| s.trim());
+            match (key, value) {
+                (Some(k), Some(v)) => Some((k, v)),
+                _ => None,
+            }
+        })
+        .collect::<HashMap<_, _>>();
+
+    parsed
+        .iter()
+        .filter_map(|(key, value)| line_to_metrics(key, value, namespace, now, tags))
+        .fold(vec![], |mut acc, v| {
+            match v {
+                Ok(metrics) => metrics.into_iter().for_each(|v| acc.push(Ok(v))),
+                Err(err) => acc.push(Err(err)),
+            };
+            acc
+        })
+        .into_iter()
+}
+
+#[derive(Debug)]
+pub struct ParseError {
+    key: String,
+    err: Box<dyn Error>,
```

Comment:
Can this have a concrete type, or are there multiple types of error handled here? If multiple types, did you consider an `enum ParseError`?

Reply:
I believe there are at least two types of parse errors, one for int and one for float. I think I did, at one point, have it as an enum, but it didn't seem worth it since all we were doing was just propagating it up without handling.

I guess, generally, do you feel it's better to be specific with the error types even if the caller doesn't care?

Reply:
Generally, I think it's better to be specific with _all_ types, even if the caller doesn't care, as it avoids invoking more "mechanism" at run time (allocations, dynamic dispatch, etc). However, it doesn't matter as much when it isn't the hot path and, by definition, exceptional errors are not the hot path. In the case of errors, though, adding a layer of wrapping adds context as to why the error happened. So it's partly a judgment call.

Reply:
:+1: that makes sense. It seems more flexible too in that if we add another caller that would care, the specificity would already be there. Thanks for the explanation.